astropy/units/equivalencies.py

```diff
@@ -561,6 +564,268 @@ def mass_energy():
+
+    def _homogeneous_conversion(
+            self, source_quantity,
+            target_unit):
+        """Composite quantity conversion. E.g. meter-cubed to second-cubed,
+        but not m/s to dimensionless units.
+
+        Parameters
+        ----------
+        source_quantity_value: float
+            source quantity value
+        source_units: list
+            list of units
+
+        """
+        assert len(source_quantity.unit.bases) == len(target_unit.bases) == 1
+        assert source_quantity.unit.powers == target_unit.powers
+
+        res = source_quantity.value
+
+        res *= self._irreducible_conversion(
+            1.0 * source_quantity.unit.bases[0],
+            target_unit.bases[0]
+        )**target_unit.powers[0]
+
+        return res
+
+
+def geometrized(physical_type):
+    """
+    Convert quantities to geometric units.
+
+    Parameters
+    ----------
+    physical_type: str or `~astropy.units.PhysicalType`
+        Physical type to convert.
+        Argument to :meth:`astropy.units.get_physical_type`
+
+    Examples
+    --------
+    >>> from astropy import constants as c, units as u
+    >>> mass = 2 * u.M_sun
+    >>> mass.to('km', equivalencies=u.geometrized('mass'))
+    <Quantity 2.95325008 km>
+    >>> time = 1 * u.microsecond
+    >>> time.to('solMass', equivalencies=u.geometrized('time'))
+    <Quantity 0.20302545 solMass>
+    >>> time.to('kg', equivalencies=u.geometrized('time'))
+    <Quantity 4.03697802e+29 kg>
+    """
+    physical_type = physical.get_physical_type(physical_type)
+    mass_length_time_physical_types = [
+        physical.length,
+        physical.mass,
+        physical.time,
+    ]
+
+    volume_physical_types = [
+        physical.volume
+    ]
+
+    area_physical_types = [
+        physical.area,
+    ]
+
+    temperature_physical_types = [
+        physical.temperature,
+    ]
+
+    velocity_physical_types = [
+        physical.velocity,
+    ]
+
+    allowed_physical_types = mass_length_time_physical_types
+    allowed_physical_types += volume_physical_types
+    allowed_physical_types += area_physical_types
+    allowed_physical_types += temperature_physical_types
+    allowed_physical_types += velocity_physical_types
+
+    assert physical_type in allowed_physical_types, (
+        "Allowed values : "
+        f"{[v._physical_type for v in allowed_physical_types]}"
+    )
+
+    conversions = GeometricUnitConversionFactors()
+
+    if physical_type in velocity_physical_types:
+
+        return Equivalency([
+            # velocity
+            (
+                si.m / si.s, Unit(),
+                lambda velocity: velocity / _si.c.value,
+                lambda v_over_c: v_over_c * _si.c.value,
+            )
+         ])
+
+    elif physical_type in area_physical_types:
+
+        return Equivalency([
+            # area-time**2
+            (
+                si.m**2, si.s**2,
+                lambda meter_sq: conversions._homogeneous_conversion(
```

Comment:
I think all these equivalencies can be defined external to this function.
And @mhvk mentioned that these might not even require lambda functions. Not sure if that's still the case, but worth checking out!

Reply:
OK I see that there is a stage where the tests are run in parallel: https://github.com/astropy/astropy/blob/main/.github/workflows/ci_workflows.yml#L150-L181 which is passing, so I believe there is no issue with serialization etc

Reply:
My apologies actually. I see that this stage had passed in a previous pipeline, but in the most recent one it has failed. Please ignore my above comment

Reply:
OK, the failure was caused by a typo in a docstring unrelated to what I had mentioned before.

I am guessing your concern is regarding the serialization of functions for multiprocessing. I believe the parallel mode for tox in the _32-bit and parallel_ stage would have caught that. But, like I mentioned, my implementation follows that of most other existing equivalencies. Maybe refactoring the callables could be an item for a later release if its an issue.

Reply:
As @nstarman noted, there should be no need for the `lambda` functions.